homeassistant/components/blebox/__init__.py

```diff
@@ -0,0 +1,105 @@
+"""The BleBox devices integration."""
+import asyncio
+import logging
+
+from blebox_uniapi.error import Error
+from blebox_uniapi.products import Products
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+CONFIG_SCHEMA = vol.Schema({DOMAIN: vol.Schema({})}, extra=vol.ALLOW_EXTRA)
+
+PLATFORMS = ["cover"]
+
+PARALLEL_UPDATES = 0
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the BleBox devices component."""
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Set up BleBox devices from a config entry."""
+
+    for component in PLATFORMS:
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, component)
+        )
+
+    return True
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry):
+    """Unload a config entry."""
+    unload_ok = all(
+        await asyncio.gather(
+            *[
+                hass.config_entries.async_forward_entry_unload(entry, platform)
+                for platform in PLATFORMS
+            ]
+        )
+    )
+
+    return unload_ok
+
+
+async def async_create_blebox_entities(
+    api_host, async_add, entity_klass, entity_type, exception
+):
+    """Create entities from a BleBox product's features."""
+    try:
+        product = await Products.async_from_host(api_host)
+    except Error as ex:
+        _LOGGER.error("Identify failed at %s:%d (%s)", api_host.host, api_host.port, ex)
+        raise exception from ex
+
+    entities = []
+    for feature in product.features[entity_type]:
+        entities.append(entity_klass(feature))
+
+    async_add(entities, True)
+    return True
+
+
+class BleBoxEntity:
+    """Implements methods common among BleBox entities."""
+
+    def __init__(self, feature):
+        """Initialize the cover entity."""
+        self._feature = feature
+
+    @property
+    def name(self):
+        """Return the name."""
+        return self._feature.full_name
+
+    @property
+    def unique_id(self):
+        """Return a unique id."""
+        return self._feature.unique_id
+
+    async def async_update(self):
+        """Update the cover state."""
+        try:
+            await self._feature.async_update()
+        except Error as ex:
+            _LOGGER.error("Updating '%s' failed: %s", self.name, ex)
```

Comment:
If the update fails it would be nice to show that its not available anymore
```
   self._available = False
```

```
    @property
    def available(self):
        """...."""
        return self._available
```

Reply:
https://developers.home-assistant.io/docs/integration_quality_scale_index/
`Handles device/service unavailable. Log a warning once when unavailable, log once when reconnected.`

The standard only speaks to handle the logging.  If you think it needs 3 update failures to declare unavailable that seems fine to me, however someone might feel differently.

Probably best to leave for now and add later in a followup as to not hold things up.

Reply:
I think it's best for the client library to inform about the availability status after every update.

(Especially since the "algorithm" for it can be device-specific and even case-specific).

Then, the Home Assistant BleBox integration would only handle logging according to the change in availability.